docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,325 @@
+
+The formatted output is not intended to supersede the prettified human-readable and friendly output text of ORAS CLI. It aims to provide a programming-friendly experience for developers who want to automate their workflows without needing to parse unstructured text.
+
+## Guidance
+
+Provide two major options to enable users to define the output format of ORAS commands:
+
+- Use `--output` to output a file or directory in the filesystem
+- Use `--format` to format the output of ORAS commands into different data formats or enable process output data using the given Go template
+  - Use `--format json|tree|table` to print the output in prettified JSON, tree view, or table view
+  - Use `--format '{{ GO_TEMPLATE_FUNCTION }}'` to enable extract or compose the output data using Go template functions 
+
+> [!NOTE]
+> - `--output -` and `--format` can not be used at the same time due to conflicts.
+
+See example use cases of formatted output for `oras manifest fetch`:
+
+- If user doesn't specify `--format` flag, the default output should be raw JSON data. This is the current behavior of `oras manifest fetch`.
+
+- Use `--output <file>` and `--format` at the same time, a manifest file should be produced in the filesystem and the `mediaType` value should be outputted on the console:
+
+```console
+$ oras manifest fetch $REGISTRY/$REPO:$TAG --output sample-manifest.json --format {{.config.MediaType}}
+application/vnd.oci.empty.v1+json
+```
+
+View the contents of the generated manifest within specified `sample-manifest.json` file. The contents should be raw JSON data:
+
+```console
+$ cat sample-manifest.json
+{"schemaVersion":2,"mediaType":"application/vnd.oci.image.manifest.v1+json","artifactType":"application/vnd.unknown.artifact.v1","config":{"mediaType":"application/vnd.oci.empty.v1+json","digest":"sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a","size":2,"data":"e30="},"layers":[{"mediaType":"application/vnd.oci.image.layer.v1.tar","digest":"sha256:6cb759c4296e67e35b0367f3c0f51dfdb776a0c99a45f39d0476e43d82696d65","size":14477,"annotations":{"org.opencontainers.image.title":"sbom.spdx"}},{"mediaType":"application/vnd.oci.image.layer.v1.tar","digest":"sha256:54c0e84503c8790e03afe34bfc05a5ce45c933430cfd9c5f8a99d2c89f1f1b69","size":6639,"annotations":{"org.opencontainers.image.title":"scan-test-verify-image.json"}}],"annotations":{"org.opencontainers.image.created":"2023-12-15T09:41:54Z"}}
+```
+
+- Use `--format json` to print the output in prettified JSON:
+
+```bash
+oras manifest fetch $REGISTRY/$REPO:$TAG --format json
+```
+
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.unknown.artifact.v1",
+  "config": {
+    "mediaType": "application/vnd.oci.empty.v1+json",
+    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
+    "size": 2,
+    "data": "e30="
+  },
+  "layers": [
```

Comment:
As we discussed in the meeting, with the current proposal, unless I have advanced knowledge or Go templates, I will not be able to implement the following scenario:
1. Use single command to only see the layers in `json` format
2. Use single command to only see the layers in `tree` format

Ideally I would like to say as a user: "Give me this subset of data in this format". What I would expect is to have two separate switches - one for what format I would like to use (`json`, `tree`, `table`, `yaml`, `txt`) and one for what data I want to see (just the layers, just the mediaType, just the digest, etc.

Reply:
1. Scenario 1 can be achieved via a single command.
2. `tree` format will be supported in `oras discover` only in this iteration and `--format {{ GO_TEMPLATE_FUNCTION }} can be used to filter out the data that users only want to see in output.